Add section on TPM attestation (#49)
hannestschofenig/tschofenig-ids: tls-attestation/draft-fossati-tls-attestation.md

   indicates the attestation type carried in the Certificate payload.
   Note that only a single value is permitted in the 
   server_attestation_type extension when carried in the EncryptedExtensions
   message.

# TPM Attestation

> One type of hardware RoT that can be used to prove control over a certain platform is the {{TPM2.0}}. TPMs natively offer interfaces for producing both key and platform attestation tokens.

Reviewer: ```suggestion
The Trusted Platform Module (TPM) {{TPM2.0}} is one type of hardware RoT. TPMs offer the ability to produce both key and platform attestation tokens.
```

Author: Done

## Platform Attestation

> Platform Configuration Registers (PCRs) represent the core mechanism in TPMs for measuring and conveying information about the platform state via remote attestation. While specifications exist for assigning individual PCRs to specific software components, the choice of which combination of PCRs to include for any attestation procedure (and which hashing algorithm to use) is left to the parties involved. The agreement over and configuration of the PCR selection falls outside the scope of this specification and thus expected to occur out-of-band.

Reviewer: ```suggestion
Platform Configuration Registers (PCRs) represent the core mechanism in TPMs for measuring and conveying information about the platform state via remote attestation. While specifications exist for assigning individual PCRs to specific software components, the choice of which combination of PCRs to include for any attestation procedure (and which hashing algorithm to use) is left to the parties involved. The agreement over and the configuration of the PCR selection falls outside the scope of this specification and is thus expected to occur out-of-band.
```

Author: Done

The attestation evidence is produced through the TPM2_Quote operation. The evidence along with all other relevant metadata is transmitted in a format derived from the {{WebAuthn}} Attestation Statements. This format and the workflows around it are defined below.

### TPM Platform Attestation Statement Format

> The TPM Platform Attestation Statement is a modified version of the TPM Attestation Statement Format which covers key attestation tokens.

Author: Done

~~~~
    tpmPlatStmtFormat = {
                          ver: "2.0",
                          (
                              alg: COSEAlgorithmIdentifier,
                              x5c: [ pakCert: bytes, * (caCert: bytes) ]
                          )
                          sig: bytes,
                          attestInfo: bytes,
                        }
~~~~
{: #figure-tpm-plat-attest-stmt title="TPM Platform Attestation Statement Format"}

- _ver_: The version of the TPM specification to which the signature conforms.
- _alg_: A COSEAlgorithmIdentifier containing the identifier of the algorithm used to generate the attestation signature.

> - _x5c_: A certificate for the PAK, followed by its certificate chain (in X.509 encoding).

Author: Excellent, yes, I will reference that.

>   - _pakCert_: The PAK certificate used for the attestation, in X.509 encoding.

Reviewer: There are different ways to encode a single X.509 certificate as well, including a PEM encoding or a DER encoding.

### Signing Procedure

Generate a signature using the operation specified in Part 3, section 18.4 of {{TPM2.0}}, using the PAK as the signing key, the out-of-band agreed-upon PCR selection. Freshness of the attestation is given by the nonce provided by the relying party. The nonce is included as qualified data to the TPM2_Quote operation, concatenated with an identifier of the platform being attested, as shown below:

  _extraData_ = _platformUuid_ || _relyingPartyNonce_

The platform identifier is a 16-bytes long UUID, with the remaining data representing the nonce. The UUID is intended to help the verifier link the platform with its expected reference values.

Set the _attestInfo_ field to the quoted PCR selection produced by the operation, and _sig_ to the signature generated above.

### Verification Procedure

The inputs to the verification procedure are as follows:

- an attestation statement in the format described above
- the nonce sent to the attester
- a database of reference values for various platforms

The steps for verifying the attestation:

> - Verify that the attestation token is valid CBOR conforming to the CTAP2 form and perform CBOR decoding on it to extract the contained fields.

Reviewer: I believe a reference to CTAP2 would be needed. Do you need everything from CTAP2 or only selected parts?

Author: I only need [this part](https://fidoalliance.org/specs/fido-v2.0-ps-20190130/fido-client-to-authenticator-protocol-v2.0-ps-20190130.html#message-encoding), relating to message encoding.

> - Verify that the attestation token is valid CBOR conforming to the CTAP2 form and perform CBOR decoding on it to extract the contained fields.

Author: Done

    * Note that the remaining fields in the "Standard Attestation Structure" (Part 1, section 31.2 of {{TPM2.0}}), i.e., _qualifiedSigner_, _clockInfo_ and _firmwareVersion_ are ignored. These fields MAY be used as an input to risk engines.

- If successful, return implementation-specific values representing attestation type AttCA and attestation trust path _x5c_.

## Key Attestation

Attesting to the provenance and properties of a key is possible through a TPM if the key resides on the TPM. The TPM2.0 key attestation mechanism used in this specification is TPM2_Certify. The workflow for generating the evidence and assessing them, as well as the format used to transport them, follows closely the TPM Attestation Statement defined in section 8.3 of {{WebAuthn}}, with one modification:

> - For both signing and verification, _attToBeSigned_ is unnecessary in our case. Its hash is replaced with the nonce coming from the relying party as the qualifying data when signing, and as the expected _extraData_ value during verification.

Reviewer: ```suggestion
- For both signing and verification, _attToBeSigned_ is unnecessary and therefor its hash is replaced with the nonce coming from the relying party as the qualifying data when signing, and as the expected _extraData_ value during verification.
```

> The key used for signing - named AIK in {{WebAuthn}} - and its certificate chain is in our case the KAK and its associated certificates. The credential (i.e., attested) key is in our case the TIK.

Reviewer: ```suggestion
The WebAuthn specification {{WebAuthn}} uses the term AIK to refer to the signing key. In this specification we use the term KAK instead. The credential (i.e., attested) key is in our case the TIK.
```

Author: Done
